core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -126,11 +125,44 @@ class KafkaApis(val requestChannel: RequestChannel,
     info("Shutdown complete.")
   }
 
+  private def checkForwarding(request: RequestChannel.Request): Unit = {
+    if (!request.header.apiKey.forwardable && request.envelopeContext.isDefined) {
+      throw new IllegalStateException("Given RPC " + request.header.apiKey + " does not support forwarding.")
+    }
+  }
+
+  private def maybeForward(request: RequestChannel.Request,
+                           handler: RequestChannel.Request => Unit): Unit = {
+    if (request.envelopeContext.isDefined && request.principalSerde.isEmpty) {
+      sendErrorResponseMaybeThrottle(request, Errors.PRINCIPAL_DESERIALIZATION_FAILURE.exception())
+    } else if (request.envelopeContext.isDefined &&
+      (!request.context.fromPrivilegedListener ||
+      !authorize(request.envelopeContext.get.brokerContext, CLUSTER_ACTION, CLUSTER, CLUSTER_NAME))
+    ) {
+      // If the designated forwarding request is not coming from a privileged listener, or
+      // it fails CLUSTER_ACTION permission, we would fail the authorization.
+      sendErrorResponseMaybeThrottle(request, Errors.CLUSTER_AUTHORIZATION_FAILED.exception())
```

Comment:
One challenge we have here is that there are two levels of errors. The current patch seems to conflate the two, which makes it confusing. I think we need a structure which allows us to separate the errors possible at the envelope level and those possible at the request level. What I'm thinking is this:

1. For cluster auth and principal serde errors, we should return the envelope error and null response body.
2. For everything else, we return envelope error NONE and just pass through whatever error is in the response.

Does that make sense?

Reply:
The question would be how the forwarding broker should do the error handling for auth & principal serde exceptions. To me we should get a vanilla error response with `UNKNOWN_SERVER_ERROR` and get back to the original client? Besides that, I think we could add a differentiation here to avoid passing the serde-type errors to the client.